### ISSUE TITLE
performance on clamp

### ISSUE THREAD

**Issue by johnnychen94** [JuliaMath/FixedPointNumbers.jl#179]
This is expected, but I think it can be optimized.

```julia
julia> @btime clamp!($(rand(N0f8, 512, 512)), 0.1, 0.9);
  813.628 μs (0 allocations: 0 bytes)

julia> @btime clamp!($(rand(N0f8, 512, 512)), 0.1N0f8, 0.9N0f8);
  628.583 μs (0 allocations: 0 bytes)

julia> @btime clamp.($(rand(N0f8, 512, 512)), 0.1, 0.9);
  691.843 μs (3 allocations: 2.00 MiB)

julia> @btime clamp.($(rand(N0f8, 512, 512)), 0.1N0f8, 0.9N0f8);
  1.146 ms (3 allocations: 256.19 KiB)
```

Ref: observed in https://github.com/JuliaImages/ImageContrastAdjustment.jl/pull/28#issuecomment-601211888

**Comment by kimikage** [comment]
I don't understand what the benchmarks above are intended for.:confused:

**Comment by johnnychen94** [comment]
I was thinking a specification like this:

```julia
```

**Comment by kimikage** [comment]
`clamp!` can be optimized, but I don't want to do it because there are many "in-place" variants.
`clamp` must not be changed as suggested. `FixedPoint` is not a pixel type (e.g. `Gray{N0f8}`).

**Comment by kimikage** [comment]
TBH, I had believed that the optimization for `clamp(x::X, lo::X, hi::X) where X<:FixedPoint` was done by the compiler (LLVM backend). I think it is worth introducing.

**Comment by johnnychen94** [comment]
Oh, I wrote this done as a note to myself and I didn't spend enough time playing with it, I'm sorry I didn't make it clear.

> This is expected, but I think it can be optimized.

I was meant to say "I believe this performance gap is expected, but I think there're merits to proactively convert `lo` and `hi` to FixedPoint type because I believe the general philosophy is to not care about storage type (it's hard to intuitive during coding), and when we try to tweak performance by converting/promoting types, there's a smell to me.

I'm catching up with my own ddls and I'm in a bit of a hurry, I'll post the update later if I find anything worth optimize.

> clamp! can be optimized, but I don't want to do it because there are many "in-place" variants.

I don't quite understand the reason behind this, can you clarify it a bit? Thanks.

**Comment by kimikage** [comment]
>> clamp! can be optimized, but I don't want to do it because there are many "in-place" variants.
>
>I don't quite understand the reasoning behind this, can you clarify it a bit?

The specialization of many "in-place" methods requires much labor and offers little benefit. 

> I believe the general philosophy is to not care about storage type

I agree with you. Therefore, I think `clamp` should be used when we can use `clamp` instead of `clamp!`. When we use `clamp!`, we must "definitely" pay attention to the storage type. The performance issues above are the result of "carelessness".:confused:

Also, if we follow the philosophy, `clamp` MUST NOT convert the return type.

**Comment by timholy** [comment]
Certainly the specialization in https://github.com/JuliaMath/FixedPointNumbers.jl/issues/179#issuecomment-601451806 makes sense. The one in https://github.com/JuliaMath/FixedPointNumbers.jl/issues/179#issuecomment-601466012 is much more dangerous, since currently `clamp(x, -Inf, Inf)` succeeds for all FixedPoint values but that definition would throw an error since `Inf` does not have a representation as FixedPoint. We could, however, define `clamp_fast` and make `@fastmath` do semantic replacement (this would presumably require a PR to Base). It might take a little work to get it to hoist the conversion for broadcasting, though.

**Comment by kimikage** [comment]
I merged PR #194. If we try to improve it further I will keep this issue open, otherwise I will close this issue.

**Comment by johnnychen94** [comment]
Thanks for working on this. #194 looks like a good-enough improvement to me.
